### ISSUE TITLE
Beta-enabled chain reaches consensus failure on binary rebuild and restart

### ISSUE THREAD

**Issue by adamewozniak** [umee-network/umee#543]
## Summary of Bug

When running the beta version of the chain locally, rebuilding the binary in beta mode, stopping the chain, and then restarting, it reaches consensus failure via an error from the `x/leverage` module : 

`ERR CONSENSUS FAILURE!!! err="-0.000002092846270928 years: negative time elapsed since last interest time" module=consensus ...`

Not sure if this is a major issue, but still worth jotting down 

## Version

Please provide the output of the following commands:

- `$ umeed version`
price-feeder/v0.1.0-c66f8961
- `$ go version`
go version go1.17.3 darwin/arm64
- `$ uname -a`
Darwin Adams-MacBook-Pro-2.local 21.3.0 Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_ARM64_T6000 arm64

## Steps to Reproduce

1. Launch beta umee network
```
UMEE_ENABLE_BETA=true starport chain serve -c ./starport.ci.beta.yml -v --reset-once
```
2. Rebuild binary with beta support
```
UMEE_ENABLE_BETA=true make install
```
3. Stop & Restart umee network 
```
UMEE_ENABLE_BETA=true starport chain serve -c ./starport.ci.beta.yml -v
```

----

## For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by alexanderbez** [comment]
Interesting. @toteki does anything stick out here?

**Comment by toteki** [comment]
We were talking about it. and the only way for this to happen is for the `LastInterestTime` in state to be less than `ctx.BlockTime().Unix()` - it comes down to the behavior of `BlockTime` and `EndBlocker` (during which `AccrueAllInterest` sets `LastInterestTime` to `ctx.BlockTime().Unix()` after accruing interest.)

The error message posted, by the way, comes out to be a 3 minute time reversal, so much longer than a block time, and might be consistent with the time the chain was stopped before relaunch. Still can't figure out why `BlockTime` would reverse though.

**Comment by RafilxTenfen** [comment]
@adamewozniak This still happen? I couldn't reproduce

**Comment by toteki** [comment]
I suspect this is just blockTime behaving weirdly when starport restarts, but we should see if we run into it during our chain upgrade tests. If that happens, it might be worth treating unexpected elapsed times differently.

**Comment by adamewozniak** [comment]
@toteki I think we should be worried about upgrading v2 -> v3, since the chain would be required to stop and restart similarly to this. Our upgrade tests should include v2.0.0 -> v2.0.1 (when we cut it), I think that would replicate this better

**Comment by toteki** [comment]
We could do that - but negative time elapsed is a panic because if it happens for any amount of time (worst case: if block time were ever zero for some reason) then unjustified interest (worst case: 50 years worth) will accrue instantly on the _next_ block with no error message.

I want to be sure that it's not a starport-only problem before considering removing the current safety. Has anyone run into the bug without starport during this week's gov upgrade testing?

**Comment by adamewozniak** [comment]
I don't believe this has been seen in a while. Please confirm / close, either @dreamcodez, @RafilxTenfen, or @toteki

**Comment by toteki** [comment]
Been a while - I suspect it's because our local testing has evolved beyond relaunching with starport (e.g. noob scripts and multinode now).

That would mean that blocktime moving backwards was just a starport edge case we don't have a problem.

**Comment by adamewozniak** [comment]
Reopening this since @RafilxTenfen has seen it while trying to do a fork of our testnet

**Comment by adamewozniak** [comment]
Closing this back up, @RafilxTenfen was able to get the fork past this just by updating the genesis file :^)

**Comment by RafilxTenfen** [comment]
Hey, I was able to reproduce this with an single node 
Here are the steps: 
- go to main on umee or `git checkout 2efa932069272c9fcbfd8ab80e02b7141f2f393b`
- build the binary `make build`
- copy the new binary to `cp ./build/umeed ../umee-infra/noob/umeed-builds/umee-main` 
- go to [this branch ](https://github.com/umee-network/umee-infra/tree/rafilx/leverage-halt) on umee-infra
- enter the noob folder
- run the script `./leverage-halt.sh umeed-builds/umeed-main`
- Look at the logs in another shell `tail -f ./data/umeetest-1.n0.log`

__Expected error__
![image](https://user-images.githubusercontent.com/17556614/172668974-51210d2f-ad53-4b40-9557-17e82a5cf57b.png)

**Comment by RafilxTenfen** [comment]
So, this is basically happening, because when we export a genesis it saves a json as the genesis state, this genesis state has the time at which the genesis was generated ex.: 
`"genesis_time":"2022-06-10T15:09:14.61567517Z"`
When loading the [tendermint state](https://github.com/tendermint/tendermint/blob/v0.34.19/node/node.go#L680-L681) it uses the genesis file when there is no state from *.db and then it sets the context `LastBlockTime` with the [genesis time](https://github.com/tendermint/tendermint/blob/v0.34.19/state/state.go#L343-L344)
`LastBlockTime:   genDoc.GenesisTime`
and then the context header time which is used by our leverage.keeper [AccurateAllInterest](https://github.com/umee-network/umee/commit/5372343a4f187864cc37f0c0d798dc7df3de16cc) is filled with the [time from the GenesisTime](https://github.com/tendermint/tendermint/blob/v0.34.19/consensus/replay.go#L313-L314) which is before than our `last_interest_time` saved in the genesis state

**Comment by RafilxTenfen** [comment]
In my opinion, tendermint should save the last block time in the state as well... 
I will test the same thing with gaia to see if their code will produce a block.time < previousblock.time

One idea to solve this would be for us to use block heigh difference instead of time, do you see a way to use this instead of the time lapsed? @toteki

**Comment by toteki** [comment]
Possible approaches:

`Option 0`: Remove the error and allow negative time to elapse. Bad because it would allow double-interest over the same time period.
`Option 1`: Fix BlockTime behavior on export. We would need to open an issue on cosmos sdk probably (we can do that anyway) but shouldn't rely on it for our fix.
`Option 2`: Use block height. This would lead to inaccuracies each block (`6.5sec` being treated as `5sec` on average) and reduce yields.
`Option 3`: Modify `AccrueAllInterest` to treat "negative time elapsed" as "zero time elapsed", aborting the function without modifying state until an `EndBlock` where `BlockTime > LastInterestTime`.

**Comment by RafilxTenfen** [comment]
> Possible approaches:
> 
> `Option 0`: Remove the error and allow negative time to elapse. Bad because it would allow double-interest over the same time period. `Option 1`: Fix BlockTime behavior on export. We would need to open an issue on cosmos sdk probably (we can do that anyway) but shouldn't rely on it for our fix. `Option 2`: Use block height. This would lead to inaccuracies each block (`6.5sec` being treated as `5sec` on average) and reduce yields. Also interest would fail to accrue over long periods (e.g. 7 days) where chain is down. `Option 3`: Modify `AccrueAllInterest` to treat "negative time elapsed" as "zero time elapsed", aborting the function without modifying state until an `EndBlock` where `BlockTime > LastInterestTime`.

I liked the `Option 3` ;D

**Comment by RafilxTenfen** [comment]
I was able to reproduce with gaia v702 
![image](https://user-images.githubusercontent.com/17556614/173385388-514dbf2b-d3cd-46a3-8ddb-07c677561056.png)

**Comment by RafilxTenfen** [comment]
Related to [#1533](https://github.com/cosmos/gaia/issues/1533)

**Comment by RafilxTenfen** [comment]
> > and then the context header time which is used by our leverage.keeper [AccurateAllInterest](https://github.com/umee-network/umee/commit/5372343a4f187864cc37f0c0d798dc7df3de16cc) is filled with the [time from the GenesisTime](https://github.com/tendermint/tendermint/blob/v0.34.19/consensus/replay.go#L313-L314) which is before than our last_interest_time saved in the genesis state
> 
> So our genesis is not consistent. Instead of adding an exception logic in `AccurateAllInterest` we should fix the genesis export:
> 
> * I don't see any issue with `BlockTime` - this is correct. It's the last block time.
> * When exporting the genesis, we should correctly set the `LastBlockTime` to the last a value from the state.
> * When importing genesis we should see if `LastBlockTime` is defined in genesis. Only if it's not defined we fallback to `genDoc.GenesisTime`

But the exporting / importing stuffs from blocks is done in tendermint/cosmos-sdk layer, right?

**Comment by brianosaurus** [comment]
I'm unsure that will do it. We're having consensus failures in other places. Can you post the stack trace and error message? As text (not an image)? I would like to see it.

**Comment by robert-zaremba** [comment]
> But the exporting / importing stuffs from blocks is done in tendermint/cosmos-sdk layer, right?

No. We control how to handle import / export. https://github.com/umee-network/umee/blob/main/x/leverage/genesis.go

**Comment by toteki** [comment]
> > But the exporting / importing stuffs from blocks is done in tendermint/cosmos-sdk layer, right?
> 
> No. We control how to handle import / export. https://github.com/umee-network/umee/blob/main/x/leverage/genesis.go

The inaccurate time isn't in leverage export - it's coming from BlockTime() itself going backwards. That's also why Rafael was able to reproduce with `gaiad`

**Comment by robert-zaremba** [comment]
oh, I thought we have our own `LastBlockTime` in leverage.

**Comment by RafilxTenfen** [comment]
> I'm unsure that will do it. We're having consensus failures in other places. Can you post the stack trace and error message? As text (not an image)? I would like to see it.
On gaiad is not going to have an consensus failure, but the next block after exported genesis happen will have the genesis block time, you can reproduce this here https://github.com/cosmos/gaia/issues/1533#issue-1269809258

You can reproduce the consensus failure, running [theses steps](https://github.com/umee-network/umee/issues/543#issuecomment-1150137223) 
but I will provide the error as text 

[Here](https://gist.github.com/RafilxTenfen/8ff8dea193ef9887e447901e4978cd16)

**Comment by robert-zaremba** [comment]
I see, I didn't look at https://github.com/cosmos/gaia/issues/1533

In that case I agree: let's:
* change log to Warn
* add todo to solve it once the bug will be resolved by Tendermint.

**Comment by RafilxTenfen** [comment]
> I see, I didn't look at [cosmos/gaia#1533](https://github.com/cosmos/gaia/issues/1533)
> 
> In that case I agree: let's:
> 
> * change log to Warn
> * add todo to solve it once the bug will be resolved by Tendermint.

https://github.com/umee-network/umee/pull/1018#issuecomment-1156778285
